annotated.go

```diff
@@ -195,6 +196,318 @@ func ResultTags(tags ...string) Annotation {
 	return resultTagsAnnotation{tags}
 }
 
+type _lifecycleHookAnnotationType int
+
+const (
+	_unknownHookType _lifecycleHookAnnotationType = iota
+	_onStartHookType
+	_onStopHookType
+)
+
+type lifecycleHookAnnotation struct {
+	Type   _lifecycleHookAnnotationType
+	Target interface{}
+}
+
+func (la *lifecycleHookAnnotation) String() string {
+	name := "UnknownHookAnnotation"
+	switch la.Type {
+	case _onStartHookType:
+		name = _onStartHook
+	case _onStopHookType:
+		name = _onStopHook
+	}
+	return name
+}
+
+func (la *lifecycleHookAnnotation) apply(ann *annotated) error {
+	if la.Target == nil {
+		return fmt.Errorf(
+			"cannot use nil function for %q hook annotation",
+			la,
+		)
+	}
+
+	for _, h := range ann.Hooks {
+		if la.Type == h.Type {
+			return fmt.Errorf(
+				"cannot apply more than one %q hook annotation",
+				la,
+			)
+		}
+	}
+
+	ft := la.targetType()
+	if ft.Kind() != reflect.Func {
+		return fmt.Errorf(
+			"must provide function for %q hook, got %v (%T)",
+			la,
+			la.Target,
+			la.Target,
+		)
+	}
+
+	if ft.NumIn() < 1 || ft.In(0) != _typeOfContext {
+		return fmt.Errorf(
+			"first argument of hook must be context.Context, got %v (%T)",
+			la.Target,
+			la.Target,
+		)
+	}
+
+	hasOut := ft.NumOut() == 1
+	returnsErr := hasOut && ft.Out(0) == _typeOfError
+
+	if !hasOut || !returnsErr {
+		return fmt.Errorf(
+			"hooks must return only an error type, got %v (%T)",
+			la.Target,
+			la.Target,
+		)
+	}
+
+	if ft.IsVariadic() {
+		return fmt.Errorf(
+			"hooks must not accept variatic parameters, got %v (%T)",
+			la.Target,
+			la.Target,
+		)
+	}
+
+	ann.Hooks = append(ann.Hooks, la)
+	return nil
+}
+
+var (
+	_typeOfLifecycle reflect.Type = reflect.TypeOf((*Lifecycle)(nil)).Elem()
+	_typeOfContext   reflect.Type = reflect.TypeOf((*context.Context)(nil)).Elem()
+)
+
+func (la *lifecycleHookAnnotation) targetType() (targetType reflect.Type) {
+	return reflect.TypeOf(la.Target)
+}
```

Comment:
do we really need this to be its own method when it's just a simple one liner function call?